File: packages/ui-android/chorus-ui/src/main/kotlin/dev/blinddsai/chorus/ui/chip/ChorusChip.kt
package dev.blinddsai.chorus.ui.chip

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Text
import androidx.compose.material.ripple.rememberRipple
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.foundation.Image
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import dev.blinddsai.chorus.tokens.ChorusTheme
import dev.blinddsai.chorus.tokens.generated.ChorusBorderWidth
import dev.blinddsai.chorus.tokens.generated.ChorusIconSize
import dev.blinddsai.chorus.tokens.generated.ChorusLayout
import dev.blinddsai.chorus.tokens.generated.ChorusPalette
import dev.blinddsai.chorus.tokens.generated.ChorusRadius
import dev.blinddsai.chorus.tokens.generated.ChorusStateOpacity
import dev.blinddsai.chorus.tokens.generated.ChorusTypography

/**
 * Compose port of the Chorus Chip family.
 *
 *  - [ChorusFilterChip] — selectable capsule; transparent + hairline
 *    when unselected, inverse-surface fill when selected.
 *  - [ChorusTagChip] — passive (or dismissable via `onDismiss`)
 *    metadata label with a translucent neutral background that adapts
 *    to whatever surface sits behind.
 */

@Composable
fun ChorusFilterChip(
    label: String,
    selected: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    leadingIcon: Painter? = null,
    trailingIcon: Painter? = null,
    enabled: Boolean = true,
) {
    val c = ChorusTheme.colors
    val background = if (selected) c.inverseSurface else Color.Transparent
    val labelColor = if (selected) c.inverseOnSurface else c.onSurface
    val border: Pair<Dp, Color>? =
        if (selected) null else ChorusBorderWidth.hairline to c.outlineVariant

    ChipShell(
        radius = ChorusRadius.full,
        minHeight = 32.dp, // ref.space.400
        paddingBlock = ChorusLayout.Container.s2xs,
        paddingInline = ChorusLayout.Container.sm,
        labelInset = ChorusLayout.Container.s2xs,
        background = background,
        labelColor = labelColor,
        border = border,
        onClick = onClick,
        enabled = enabled,
        modifier = modifier,
    ) {
        if (leadingIcon != null) {
            Image(painter = leadingIcon, contentDescription = null,
                  modifier = Modifier.size(ChorusIconSize.md))
        }
        Text(label, style = ChorusTypography.Label.sm, color = labelColor, maxLines = 1)
        if (trailingIcon != null) {
            Image(painter = trailingIcon, contentDescription = null,
                  modifier = Modifier.size(ChorusIconSize.md))
        }
    }
}

@Composable
fun ChorusTagChip(
    label: String,
    modifier: Modifier = Modifier,
    onDismiss: (() -> Unit)? = null,
    trailingIcon: Painter? = null,
    enabled: Boolean = true,
) {
    val c = ChorusTheme.colors
    val dark = isSystemInDarkTheme()
    val background = if (dark) ChorusPalette.White.s200 else ChorusPalette.Black.s200

    ChipShell(
        radius = ChorusRadius.sm,
        minHeight = 24.dp, // ref.space.300
        paddingBlock = ChorusLayout.Container.s2xs,
        paddingInline = ChorusLayout.Container.xs,
        labelInset = ChorusLayout.Container.s2xs,
        background = background,
        labelColor = c.onSurface,
        border = null,
        onClick = onDismiss,
        enabled = enabled,
        modifier = modifier,
    ) {
        Text(label, style = ChorusTypography.Label.sm, color = c.onSurface, maxLines = 1)
        if (trailingIcon != null) {
            Image(painter = trailingIcon, contentDescription = null,
                  modifier = Modifier.size(ChorusIconSize.md))
        }
    }
}

@Composable
private fun ChipShell(
    radius: Dp,
    minHeight: Dp,
    paddingBlock: Dp,
    paddingInline: Dp,
    labelInset: Dp,
    background: Color,
    labelColor: Color,
    border: Pair<Dp, Color>?,
    onClick: (() -> Unit)?,
    enabled: Boolean,
    modifier: Modifier,
    content: @Composable () -> Unit,
) {
    val shape = RoundedCornerShape(radius)
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()

    val clickModifier = if (onClick != null) {
        Modifier.clickable(
            interactionSource = interactionSource,
            indication = rememberRipple(color = labelColor),
            enabled = enabled,
            role = Role.Button,
            onClick = onClick,
        )
    } else Modifier

    Box(
        modifier = modifier
            .defaultMinSize(minHeight = minHeight)
            .alpha(if (enabled) 1f else ChorusStateOpacity.disabled)
            .clip(shape)
            .background(background, shape)
            .then(border?.let { (w, c) -> Modifier.border(BorderStroke(w, c), shape) } ?: Modifier)
            .then(clickModifier)
            .padding(PaddingValues(horizontal = paddingInline, vertical = paddingBlock)),
        contentAlignment = Alignment.Center,
    ) {
        if (pressed && onClick != null) {
            Box(
                modifier = Modifier
                    .matchParentSize()
                    .background(labelColor.copy(alpha = ChorusStateOpacity.pressed))
            )
        }
        Row(
            horizontalArrangement = Arrangement.spacedBy(labelInset, Alignment.CenterHorizontally),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            content()
        }
    }
}

File: packages/ui-ios/Sources/ChorusUI/Chip/ChorusChip.swift
import SwiftUI
import ChorusTokens

/// SwiftUI port of the Chorus Chip family.
/// Specs: `schema/components/chip/filter.spec.json`, `.../tag.spec.json`.

public enum ChorusChipVariant { case filter, tag }

public struct ChorusChip: View {
    private let label: String
    private let variant: ChorusChipVariant
    private let selected: Bool
    private let leadingIcon: Image?
    private let trailingIcon: Image?
    private let action: (() -> Void)?

    @Environment(\.colorScheme) private var colorScheme
    @Environment(\.isEnabled) private var isEnabled

    public static func filter(
        _ label: String,
        selected: Bool,
        leadingIcon: Image? = nil,
        trailingIcon: Image? = nil,
        action: @escaping () -> Void
    ) -> ChorusChip {
        ChorusChip(
            label: label,
            variant: .filter,
            selected: selected,
            leadingIcon: leadingIcon,
            trailingIcon: trailingIcon,
            action: action
        )
    }

    /// Tag chip — passive (omit `onDismiss`) or interactive (pass an
    /// `onDismiss` to render the × dismiss glyph and make the chip
    /// keyboard-focusable).
    public static func tag(
        _ label: String,
        onDismiss: (() -> Void)? = nil
    ) -> ChorusChip {
        ChorusChip(
            label: label,
            variant: .tag,
            selected: false,
            leadingIcon: nil,
            trailingIcon: onDismiss != nil ? Image(systemName: "xmark") : nil,
            action: onDismiss
        )
    }

    private init(
        label: String,
        variant: ChorusChipVariant,
        selected: Bool,
        leadingIcon: Image?,
        trailingIcon: Image?,
        action: (() -> Void)?
    ) {
        self.label = label
        self.variant = variant
        self.selected = selected
        self.leadingIcon = leadingIcon
        self.trailingIcon = trailingIcon
        self.action = action
    }

    public var body: some View {
        let s = sizing()
        let surface = surface()

        let content = HStack(spacing: 0) {
            if let icon = leadingIcon {
                icon
                    .resizable()
                    .renderingMode(.template)
                    .frame(width: s.iconSize, height: s.iconSize)
            }
            Text(label)
                .lineLimit(1)
                .padding(.horizontal, s.labelInset)
            if let icon = trailingIcon {
                icon
                    .resizable()
                    .renderingMode(.template)
                    .frame(width: s.iconSize, height: s.iconSize)
            }
        }
        .frame(minHeight: s.minHeight)
        .padding(.vertical, s.paddingBlock)
        .padding(.horizontal, s.paddingInline)

        Group {
            if let action {
                Button(action: action) { content }
                    .buttonStyle(ChorusSurfaceButtonStyle(surface: surface, radius: s.radius))
            } else {
                content
                    .background(surface.background)
                    .clipShape(RoundedRectangle(cornerRadius: s.radius, style: .continuous))
                    .foregroundStyle(surface.label)
            }
        }
        .font(.system(size: s.labelTypo.size, weight: s.labelTypo.fontWeight))
        .tracking(s.labelTypo.trackingPoints)
        .opacity(isEnabled ? 1.0 : ChorusStateOpacity.disabled)
    }

    private struct Sizing {
        let paddingBlock: CGFloat
        let paddingInline: CGFloat
        let labelInset: CGFloat
        let minHeight: CGFloat
        let radius: CGFloat
        let iconSize: CGFloat
        let labelTypo: ChorusTextStyle
    }

    private func sizing() -> Sizing {
        switch variant {
        case .filter:
            return Sizing(
                paddingBlock: ChorusLayout.Container.s2xs,
                paddingInline: ChorusLayout.Container.sm,
                labelInset: ChorusLayout.Container.s2xs,
                minHeight: 32,
                radius: ChorusRadius.full,
                iconSize: ChorusIconSize.md,
                labelTypo: ChorusTypography.Label.sm
            )
        case .tag:
            return Sizing(
                paddingBlock: ChorusLayout.Container.s2xs,
                paddingInline: ChorusLayout.Container.xs,
                labelInset: ChorusLayout.Container.s2xs,
                minHeight: 24,
                radius: ChorusRadius.sm,
                iconSize: ChorusIconSize.md,
                labelTypo: ChorusTypography.Label.sm
            )
        }
    }

    private func surface() -> ChorusSurface {
        let p = ChorusColors.themed(for: colorScheme)
        switch variant {
        case .filter:
            if selected {
                return ChorusSurface(background: p.inverseSurface, label: p.inverseOnSurface, border: nil)
            }
            return ChorusSurface(
                background: .clear,
                label: p.onSurface,
                border: (ChorusBorderWidth.hairline, p.outlineVariant)
            )
        case .tag:
            // Tag's translucent neutral has to flip across themes so the
            // chip adapts to whichever surface sits behind it; the rest
            // of the palette is theme-resolved already.
            let bg = colorScheme == .dark
                ? ChorusPalette.White.s200
                : ChorusPalette.Black.s200
            return ChorusSurface(background: bg, label: p.onSurface, border: nil)
        }
    }
}
